README.markdown

```diff
@@ -122,11 +179,87 @@ An array of services to start, normally `puppet`. If the array is empty, no serv
   service_names => ['puppet']
 ```
 
+
 ##### `source`
 
-Alternate source from which you wish to download the latest version of Puppet. On the Windows operating system this is the absolute path to the MSI file to install, for example:
+INCLUDED FOR COMPATIBILITY WITH MODULE VERSIONS 1.0/2.0. PREFER USE OF "absolute_source", "(yum/apt/mac etc.)_source", "alternate_pe_source" OVER USE OF "source".
+
+The location to find packages. Replaces base URL for unix/MacOS agents, used as fully qualified path in windows.
+
+Unix/MacOS
+``` puppet
+  source => 'https://alternate-pe-master.com:8140'
+```
+
+Windows
+``` puppet
+  source => 'C:/packages/puppet-agent-6.0.0-x64.msi'
+```
+
+##### `absolute_source`
+
+Absolute ("fully qualified") source path from which you wish to download the latest version of Puppet. No path structure or package name is assumed: the fully qualified path to the package itself must be provided.
+``` puppet
+  absolute_source => 'C:/packages/puppet-agent-6.0.0-x64.msi'
+```
+
+##### `yum_source`
+
+Base URL of a location or mirrors of yum.puppet.com downloads sites. Directories under the URL should match the structure of the yum.puppet.com
+``` puppet
+  yum_source => 'https://my-puppet-yum-mirror.com'
+```
+
+##### `apt_source`
+
+Base URL of a location or mirrors of apt.puppet.com downloads sites. Directories under the URL should match the structure of the apt.puppet.com
+``` puppet
+  apt_source => 'https://my-puppet-apt-mirror.com'
+```
+
+##### `mac_source`
+
+Base URL of a location or mirrors of downloads.puppet.com downloads site that serves MacOS packages. Directories under the URL should match the structure of the downloads.puppet.com site
+``` puppet
+  mac_source => 'https://my-puppet-downloads-mirror.com'
+```
+
+##### `windows_source`
+
+URL of a location or mirrors of downloads.puppet.com downloads site that serves packages. Directories under the URL should match the structure of downloads.puppet.com site
+``` puppet
+  windows_source => 'https://my-puppet-downloads-mirror.com'
+```
+
+##### `solaris_source`
+
+Base URL of the location of a mirror for Solaris packages. Currently, solaris packages can only be made available by using puppetlabs-pe_repo. This means the mirror must be of a PE master package serve.
+``` puppet
+  solaris_source => 'https://my-pe_master-mirror.com'
+```
+
+##### `aix_source`
+
+Base URL of the location of a mirror for AIX packages. Currently, AIX packages can only be made available by using puppetlabs-pe_repo. This means the mirror must be of a PE master package serve.
+``` puppet
+  solaris_source => 'https://my-pe_master-mirror.com'
+```
+
+
```

Comment:
I see a lot of "must match structure" types of comments. I'm assuming we can't readily detect this and just do the right thing?

Reply:
No not really.

It's probably _technically_ possible, but a huge undertaking and certainly not in the scope of this module. 

What we want to support is direct mirrors of our sources, which honestly is probably all anyone needs.